### ISSUE TITLE
Enhancement: Add make hostchk rule

### ISSUE THREAD

**Issue by lcn2** [ioccc-src/mkiocccentry#250]
The `make hostchk` rule performs some basic checks to see if the host can support this repo.

The actions of `make hostchk` should be does via a `hostchk.sh` shell script.

We do **NOT** need to test everything about the host, just a few basic elements.

For example, test if `tar` as the required flags such as `-J` and `--format=v7`.

If a problem is detected, a suggestion could be given to the user.  For example, if they are running under an OS that does not have a modern enough `tar` (such as perhaps `IRIX` or maybe an old `OpenBSD`),  one should recommend they download a install a more up to date `tar such as [GNU Tar](https://www.gnu.org/software/tar/).

**Comment by lcn2** [comment]
The output of this rule could be included in the `bug-report.txt` file.  See issue #248

**Comment by xexyl** [comment]
> The `make hostchk` rule performs some basic checks to see if the host can support this repo.
> 
> The actions of `make hostchk` should be does via a `hostchk.sh` shell script.
> 
> We do **NOT** need to test everything about the host, just a few basic elements.
> 
> For example, test if `tar` as the required flags such as `-J` and `--format=v7`.
> 
> If a problem is detected, a suggestion could be given to the user. For example, if they are running under an OS that does not have a modern enough `tar` (such as perhaps `IRIX` or maybe an old `OpenBSD`), one should recommend they download a install a more up to date `tar` such as [GNU Tar](https://www.gnu.org/software/tar/).

As far as `OpenBSD` it appears that it doesn't support `-J` for example so this would have to be done anyway. At least according to the man page.

Of course compiling tar might be too much for some people even who participate in the contest. Maybe not. I don't know. I have fond memories of doing this years ago but I don't miss it either for production systems. However helping them do this would be out of the realm of the repo I think and it's probable that most people that are participating do have a tar that works. Most that have such systems would probably have experience anyway as they'd need it - and maybe they're actually making sure their entry works on other older systems.

But what I'm trying to say I guess is that we can of course only do so much when we require certain features be there. Whether this is an issue or not I don't know.

**Comment by xexyl** [comment]
..and I have to leave now. I'll bbl. Good day!

**Comment by lcn2** [comment]
> > The `make hostchk` rule performs some basic checks to see if the host can support this repo.
> > The actions of `make hostchk` should be does via a `hostchk.sh` shell script.
> > We do **NOT** need to test everything about the host, just a few basic elements.
> > For example, test if `tar` as the required flags such as `-J` and `--format=v7`.
> > If a problem is detected, a suggestion could be given to the user. For example, if they are running under an OS that does not have a modern enough `tar` (such as perhaps `IRIX` or maybe an old `OpenBSD`), one should recommend they download a install a more up to date `tar` such as [GNU Tar](https://www.gnu.org/software/tar/).
> 
> As far as `OpenBSD` it appears that it doesn't support `-J` for example so this would have to be done anyway. At least according to the man page.
> 
> Of course compiling tar might be too much for some people even who participate in the contest. Maybe not. I don't know. I have fond memories of doing this years ago but I don't miss it either for production systems. However helping them do this would be out of the realm of the repo I think and it's probable that most people that are participating do have a tar that works. Most that have such systems would probably have experience anyway as they'd need it - and maybe they're actually making sure their entry works on other older systems.
> 
> But what I'm trying to say I guess is that we can of course only do so much when we require certain features be there. Whether this is an issue or not I don't know.

We don't think that if `OpenBSD` has an issue, we need to go out of our way to accommodate such systems.  While not as far back as `IRIX`, `OpenBSD` tends to lag considerably behind the code curve.

We need to keep in mind that all a user needs to do is to find a suitable system on which to compile and run this code.  They can develop their entry anywhere.  They just need to find a suitable system on which to package their entry .. or risk doing it by hand :-)

We will spare you comments about interactions with the "maintainer" of `OpenBSD` as this is OT.

There are plenty other distributions, even BSD systems, on which the repo should work.   Anyone who buys into the `OpenBSD` is secure and is using such systems has had plenty of times where what they want to do is blocked by the fact that `OpenBSD` is so back-rev.  This repo not compiling `OpenBSD` will not come as a surprise to any long time `OpenBSD` user.

We also go to some lengths to not force `tar` to be anywhere in particular by the `-t tar` flag.  So such a `OpenBSD` user just has to compile tar somewhere and use the `-t tar` flag.

About the best one might do for such corer case systems such as `OpenBSD` or `IRIX` is to detect the problem early and report the problem during `make hostchk`.

**Comment by xexyl** [comment]
> > > The `make hostchk` rule performs some basic checks to see if the host can support this repo.
> > > The actions of `make hostchk` should be does via a `hostchk.sh` shell script.
> > > We do **NOT** need to test everything about the host, just a few basic elements.
> > > For example, test if `tar` as the required flags such as `-J` and `--format=v7`.
> > > If a problem is detected, a suggestion could be given to the user. For example, if they are running under an OS that does not have a modern enough `tar` (such as perhaps `IRIX` or maybe an old `OpenBSD`), one should recommend they download a install a more up to date `tar` such as [GNU Tar](https://www.gnu.org/software/tar/).
> > 
> > 
> > As far as `OpenBSD` it appears that it doesn't support `-J` for example so this would have to be done anyway. At least according to the man page.
> > Of course compiling tar might be too much for some people even who participate in the contest. Maybe not. I don't know. I have fond memories of doing this years ago but I don't miss it either for production systems. However helping them do this would be out of the realm of the repo I think and it's probable that most people that are participating do have a tar that works. Most that have such systems would probably have experience anyway as they'd need it - and maybe they're actually making sure their entry works on other older systems.
> > But what I'm trying to say I guess is that we can of course only do so much when we require certain features be there. Whether this is an issue or not I don't know.
> 
> We don't think that if `OpenBSD` has an issue, we need to go out of our way to accommodate such systems. While not as far back as `IRIX`, `OpenBSD` tends to lag considerably behind the code curve.
> 
Good point.

> We need to keep in mind that all a user needs to do is to find a suitable system on which to compile and run this code. They can develop their entry anywhere. They just need to find a suitable system on which to package their entry .. or risk doing it by hand :-)


More good points.

> 
> We will spare you comments about interactions with the "maintainer" of `OpenBSD` as this is OT.
> 

Oh but please do! Maybe in the other thread? As for HIM I have vague memories of him. I don't want to say that there was arrogance but I have this vague memory of some row he had with some people anyway. Maybe all of his statements were justified. I don't know and I don't care to speculate (speculation never helps anyone feel good and very often it's wrong anyway).

> There are plenty other distributions, even BSD systems, on which the repo should work. Anyone who buys into the `OpenBSD` is secure and is using such systems has had plenty of times where what they want to do is blocked by the fact that `OpenBSD` is so back-rev. This repo not compiling `OpenBSD` will not come as a surprise to any long time `OpenBSD` user.
> 

True. And even CentOS has the functionality to work and they stay behind for stability. Though they'll be staying behind for ever in 2024 due to those bloody .. well I have nothing nice to say about them ... from Red Hat deciding CentOS will die. I'm still very unhappy about that. I have to work out what to do and I'll be probably making the transition middle of next year as I don't want to run into no updates. 

This reminds me though. I have access to a Rocky linux system so I can test it under that too. 

> We also go to some lengths to not force `tar` to be anywhere in particular by the `-t tar` flag. So such a `OpenBSD` user just has to compile tar somewhere and use the `-t tar` flag.

That's true.

> 
> About the best one might do for such corer case systems such as `OpenBSD` or `IRIX` is to detect the problem early and report the problem during `make hostchk`.

Makes sense.

**Comment by xexyl** [comment]
Please see comment https://github.com/ioccc-src/mkiocccentry/issues/248#issuecomment-1158034077.

**Comment by xexyl** [comment]
> The `make hostchk` rule performs some basic checks to see if the host can support this repo.
> 
> The actions of `make hostchk` should be does via a `hostchk.sh` shell script.
> 
> We do **NOT** need to test everything about the host, just a few basic elements.
> 
> For example, test if `tar` as the required flags such as `-J` and `--format=v7`.
> 
> If a problem is detected, a suggestion could be given to the user. For example, if they are running under an OS that does not have a modern enough `tar` (such as perhaps `IRIX` or maybe an old `OpenBSD`), one should recommend they download a install a more up to date `tar` such as [GNU Tar](https://www.gnu.org/software/tar/).



> The output of this rule could be included in the `bug-report.txt` file. See issue #248



So with these it just seems that what `hostchk.sh` currently has can be simply concatenated to the bug report log, right? Is there anything that should be added or changed with this script? If not should I mark it as complete and write a man page for it ?

Is there anything already in this script that s should not be included directly in `bug-report.sh` ?

Thank you. I'll be doing something else now but maybe tomorrow I can work on this - depends on how I feel of course. Last night was truly awful so I'm happy I managed to do a bit earlier on. Maybe I can do more later but I think it's probably better that I wait on discussion before I go ahead and do these things - and it'll give me time to take it easy.

**Comment by lcn2** [comment]
See [comment 1276855496](https://github.com/ioccc-src/mkiocccentry/issues/248#issuecomment-1276855496) and [comment 1276865756](https://github.com/ioccc-src/mkiocccentry/issues/248#issuecomment-1276865756) for comments about `hostchk.sh` and Makefile.

If there is something that is still not clear, or if there is an important question you asked previously that has not been successfully addressed, please ask again, one question per comment below.

**Comment by lcn2** [comment]
See commit https://github.com/ioccc-src/mkiocccentry/commit/89b0c18cc966e42fa86b5548c333807d25a5c0c2 for an initial implementation for https://github.com/ioccc-src/mkiocccentry/issues/248#issuecomment-1276865756.

**Comment by lcn2** [comment]
Changed issue #250 to "Enhancement: write the hostchk.sh code".

There is a `hostchk` rule in the above mentioned commit.

**Comment by lcn2** [comment]
The `hostchk.sh` code, as it is now without using `-f` can be sped up by doing a _all requited system include files_ in one compile when using `-f`.

The current way, without `-f`, where system include files are tested one at a time, is fine .. especially if `-v 1` prints something before each compiler trial.

An effort should be made to have `hostchk.sh -f` be both useful and fast.  Right now it just does an `exit 0`, so the current `-f` code is NOT useful.  :-)

**Comment by xexyl** [comment]
> The `hostchk.sh` code, as it is now without using `-f` can be sped up by doing a _all requited system include files_ in one compile when using `-f`.
> 
That's true.

> The current way, without `-f`, where system include files are tested one at a time, is fine .. especially if `-v 1` prints something before each compiler trial.

I suppose so.

> 
> An effort should be made to have `hostchk.sh -f` be both useful and fast. Right now it just does an `exit 0`, so the current `-f` code is NOT useful. :-)

True. I found it very funny how you put it though - that it's so fast it does nothing. Made me laugh so thank you for that!

**Comment by xexyl** [comment]
... just did this. Not sure if I should do a commit as I'm not sure about the pull request I have - whether the sample hook is good to have or not. However if necessary I'll just reform the pull request with the same commits so I think I will. This was quite easy to do. I'll make a commit and push it and then change the title of the pull request. After that I'll look at the other thread but I'm not sure if I'll do any other script changes. We shall see.

**Comment by xexyl** [comment]
With commit https://github.com/ioccc-src/mkiocccentry/pull/362/commits/9bc8ec13e1b4eafe56e02309d1e2da39000c3d77 the question now becomes: is there anything else that should be added to this script? If not (and even if there is) the verbose levels should be checked and the messages might be improved some too.

We can discuss these here and I can make changes. If however the commit https://github.com/ioccc-src/mkiocccentry/pull/362/commits/d9007f4fc4cd8b0c39ac7ae7ec99b3886f9fbf3e is not desired and you cannot not merge it as well I will reform the pull request and then update this comment. 

I will now look at the bug-report.sh thread. I actually have some other commits as well but since I don't know about the commit I just referred to I'm going to hold off on those.

**Comment by lcn2** [comment]
> With commit [9bc8ec1](https://github.com/ioccc-src/mkiocccentry/commit/9bc8ec13e1b4eafe56e02309d1e2da39000c3d77) the question now becomes: is there anything else that should be added to this script? If not (and even if there is) the verbose levels should be checked and the messages might be improved some too.
> 
> We can discuss these here and I can make changes. If however the commit [d9007f4](https://github.com/ioccc-src/mkiocccentry/commit/d9007f4fc4cd8b0c39ac7ae7ec99b3886f9fbf3e) is not desired and you cannot not merge it as well I will reform the pull request and then update this comment.
> 
> I will now look at the bug-report.sh thread. I actually have some other commits as well but since I don't know about the commit I just referred to I'm going to hold off on those.

Initial check looks good, thanks @xexyl

**Comment by xexyl** [comment]
> > With commit [9bc8ec1](https://github.com/ioccc-src/mkiocccentry/commit/9bc8ec13e1b4eafe56e02309d1e2da39000c3d77) the question now becomes: is there anything else that should be added to this script? If not (and even if there is) the verbose levels should be checked and the messages might be improved some too.
> > We can discuss these here and I can make changes. If however the commit [d9007f4](https://github.com/ioccc-src/mkiocccentry/commit/d9007f4fc4cd8b0c39ac7ae7ec99b3886f9fbf3e) is not desired and you cannot not merge it as well I will reform the pull request and then update this comment.
> > I will now look at the bug-report.sh thread. I actually have some other commits as well but since I don't know about the commit I just referred to I'm going to hold off on those.
> 
> Initial check looks good, thanks @xexyl

Thank you. But as noted in the other thread of course more needs to be done with this script. It's good it looks good so far though.

**Comment by xexyl** [comment]
I have some commits almost ready but first I'm taking a break. Will get to it later this morning. Maybe I can start the man pages for these two scripts too but we'll see. I'll try and remember to update this comment when the commits are in place.

**Comment by xexyl** [comment]
With commits https://github.com/ioccc-src/mkiocccentry/pull/378/commits/28c57d64ba3bb9e1e20f47f6622f242131713df2 and https://github.com/ioccc-src/mkiocccentry/pull/378/commits/e1ad4f2bac13e13e64066edf377e0e82b73691ce

is this issue also resolved ? If not what needs to be changed or added? If it's good you can close it if you like - or else we can discuss whether it should be closed. Going to do other things now and I doubt I'll work on the repo more today - maybe reply to comments depending on the time any are made.

Hope you had a great sleep!

**Comment by lcn2** [comment]
We will review this tool in the near future.

See also [comment 1294147205](https://github.com/ioccc-src/mkiocccentry/issues/171#issuecomment-1294147205).

**Comment by xexyl** [comment]
> We will review this tool in the near future.
> 
> See also [comment 1294147205](https://github.com/ioccc-src/mkiocccentry/issues/171#issuecomment-1294147205).

Thank you for the update! I look forward to discussing it with you! I also hope we’re getting to the point where we might be able to run the mock contest. But in some ways I will be sad because I have been enjoying working with you so very much and I will miss it! But then I guess I will also get to work with you on the json parser repo! I also look forward to seeing what you do with the submitted entries for the mock contest! 

Otoh I also feel like because I will be working with the json parser repo I will have less time to actually work on real entries for the real contest and unfortunately I only have one entry that I haven’t submitted yet! But then if I am honest the fact I wrote a tool for you (txzchk) and co-developed json parser with you and also the other tools (plus the scripts I wrote for the contest) plus all the wonderful wonderful compliments from you (plus the wonderful facts about C and Unix history and the hilarious stories about H2G2 with Google - and everything else) .. I feel like I have won several additional entries! It’s a huge huge honour and privilege! So thank you for all of this! Means so much to me! Truly. 

I will be going to sleep soon. Please do keep me updated on chkentry and I would like to know what your birthday is if you’re okay saying when it is!

**Comment by lcn2** [comment]
> Thank you for the update! I look forward to discussing it with you! I also hope we’re getting to the point where we might be able to run the mock contest. But in some ways I will be sad because I have been enjoying working with you so very much and I will miss it! But then I guess I will also get to work with you on the json parser repo! I also look forward to seeing what you do with the submitted entries for the mock contest!

The _IOCCCMOCK_ won't be able to start until the _submit server_ is established.  That work, which will unfortunately involve the use of a python based server, has been only prototyped in a repo that is not yet been made public.

There will be some tools related to the _submit server_ that will need to be written as well.

And before  _IOCCCMOCK_ is held (and before _submit server_ is ready), there will be inviting the general public to review the code in this repo.   The code in this repo will need to be adjusted (NOT NOW) to indicate that the _submit server_ is not yet ready to test.  During this period of public review, the `bug_report.sh` tool will be tested by the public to find any issues they may find with this repo.

And then there is the massive rewrite of the [IOCCC web site](https://www.ioccc.org), using a forked website repo to experiment on, well before the main [IOCCC web site](https://www.ioccc.org) is updated.  There are some JSON files that will be part of the new web site that describe a winning author (sort of like this repo's .author.json, but with more stuff about IOCCC winners) AND describe a winning entry (sort of like this repo's .info.json, but with more stuff about IOCCC web site).  There will need to be be tools that use those JSON files to create links within the web site.  And of course, there will be PLENTY of stuff to fix in the IOCCC entry descriptions and Makefiles, etc.  We have been largely ignoring pull requests for the [IOCCC web site](https://www.ioccc.org) because we want to focus on the forked website repo.  We are in the middle of a major edit pass over the web site that has been stalled pending `chkentry(1)` passing tests: until that happens the forked website repo remains idle and it not ready for others to contribute too at this time.   The massive rewrite of the [IOCCC web site](https://www.ioccc.org) is not related to the _IOCCCMOCK_ event, but might be nice to have in better shape prior to the _IOCCCMOCK_ event starting.  So perhaps we will put effort into getting the forked website repo ready for input after `chkentry(1)` passes tests and before _submit server_ work.

We see you being able to contribute to most of those efforts mentioned above.  So you see, there is **PLENTY** of stuff to do beyond this repo going _code complete_.

**Comment by xexyl** [comment]
> > Thank you for the update! I look forward to discussing it with you! I also hope we’re getting to the point where we might be able to run the mock contest. But in some ways I will be sad because I have been enjoying working with you so very much and I will miss it! But then I guess I will also get to work with you on the json parser repo! I also look forward to seeing what you do with the submitted entries for the mock contest!
> 

First of all I hope you're sleeping well! Now for my reply to your reply.

> The _IOCCCMOCK_ won't be able to start until the _submit server_ is established. That work, which will unfortunately involve the use of a python based server, has been only prototyped in a repo that is not yet been made public.

Icky .. python. I may vomit! :-(  That's why the thumbs down. :-) On the other hand it's better than php :-) 

> 
> There will be some tools related to the _submit server_ that will need to be written as well.
> 
> And before _IOCCCMOCK_ is held (and before _submit server_ is ready), there will be inviting the general public to review the code in this repo. The code in this repo will need to be adjusted (NOT NOW) to indicate that the _submit server_ is not yet ready to test. During this period of public review, the `bug_report.sh` tool will be tested by the public to find any issues they may find with this repo.

I imagine that the general public might have some things to say yes. But I guess we'll find out in time. Great idea to have the public use the `bug_report.sh` script to find issues with the repo in portability etc. It'd be nice if some people who had access to BSD systems could give some feedback as well.  I've not used BSD (was FreeBSD) in years and as I recall I didn't really end up liking it in the end. At first I was okay with it but the primitive (compared to GNU's) libc was not nice to work with. 

> 
> And then there is the massive rewrite of the [IOCCC web site](https://www.ioccc.org), using a forked website repo to experiment on, well before the main [IOCCC web site](https://www.ioccc.org) is updated. There are some JSON files that will be part of the new web site that describe a winning author (sort of like this repo's .author.json, but with more stuff about IOCCC winners) AND describe a winning entry (sort of like this repo's .info.json, but with more stuff about IOCCC web site). There will need to be be tools that use those JSON files to create links within the web site. And of course, there will be PLENTY of stuff to fix in the IOCCC entry descriptions and Makefiles, etc. We have been largely ignoring pull requests for the [IOCCC web site](https://www.ioccc.org) because we want to focus on the forked website repo. We are in the middle of a major edit pass over the web site that has been stalled pending `chkentry(1)` passing tests: until that happens the forked website repo remains idle and it not ready for others to contribute too at this time. The massive rewrite of the [IOCCC web site](https://www.ioccc.org) is not related to the _IOCCCMOCK_ event, but might be nice to have in better shape prior to the _IOCCCMOCK_ event starting. So perhaps we will put effort into getting the forked website repo ready for input after `chkentry(1)` passes tests and before _submit server_ work.
> 
Of course. On the subject of the `.author.json` file I was wondering again what you thought about making it so users could submit more than one URL. As you know there are two I'd like to have and right now I can only submit one so I'm never sure which one to submit: I have in the past selected the `ioccc` subdomain.


> We see you being able to contribute to most of those efforts mentioned above. So you see, there is **PLENTY** of stuff to do beyond this repo going _code complete_.

Thank you for that! That as well as the rest of it (except the python part) is why the thumbs up! :-)

Good information. I certainly have more I could say about some of it but I'll not worry about it for now as I'll get a chance to do work with it later and this way it saves you time. Of course in the end when all of this is done that's when I'll miss the working with you. Still it's a huge honour and privilege so thank you again!

**Comment by lcn2** [comment]
See commit 5035a6e64a4cd0b8d1150d2ca5e0e3021657c425

**Comment by xexyl** [comment]
> The tool is very good, @xexyl .

Thank you. I'll update the comments. That being said see below.

> 
> See commit [5035a6e](https://github.com/ioccc-src/mkiocccentry/commit/5035a6e64a4cd0b8d1150d2ca5e0e3021657c425)
> 

All good except the thing you fixed that I did not think about very possibly due to being tired! I'm glad you caught it. I refer of course to:

    Fixed the potentially dangerous pre-removal the working directory
    as use of things like -w /tmp or -w . might, or parallel makes
    that have bad side effects.

Great catch and thank you for fixing that!

> ### UPDATE 0
> If all is well, we can close this enhancement issue.

Yes it can be closed if you think so. I'll update comments in a bit I hope to say it's complete and such.

**Comment by lcn2** [comment]
> Yes it can be closed if you think so. I'll update comments in a bit I hope to say it's complete and such.

Let us know when you are done and ready to close this issue.

**Comment by xexyl** [comment]
> > Yes it can be closed if you think so. I'll update comments in a bit I hope to say it's complete and such.
> 
> Let us know when you are done and ready to close this issue.

Go ahead .. and then get some sleep! I'm working on bug_report.sh. I will say it there too but you can close it: I'm just cleaning up some comments a bit.

**Comment by lcn2** [comment]
Thank you very much, @xexyl, for  your excellent work on this issue!

**Comment by xexyl** [comment]
> Thank you very much, @xexyl, for your excellent work on this issue!

Thank you very much Landon! I really needed to hear that!

**Comment by xexyl** [comment]
See https://github.com/ioccc-src/mkiocccentry/pull/389/commits/cee7f48c6472c7b2b19fe8b474d872063bd2f026.

**Comment by xexyl** [comment]
.. sigh. I have to re-fork my repo again because for some reason:

```
$ git-pull-mkiocccentry 
M	hostchk.sh
Already on 'master'
Your branch is up to date with 'origin/master'.
From https://github.com/ioccc-src/mkiocccentry
 * branch            master     -> FETCH_HEAD
Your local changes to the following files would be overwritten by merge:
  hostchk.sh
$ git diff hostchk.sh
$
```

**\*mutters something about GitHub ...\***

When I've fixed that I'll look at the JSON comment that I haven't addressed yet. That way I can look at any problems you brought up.

**Comment by xexyl** [comment]
> .. sigh. I have to re-fork my repo again because for some reason:
> 
> ```
> $ git-pull-mkiocccentry 
> M	hostchk.sh
> Already on 'master'
> Your branch is up to date with 'origin/master'.
> From https://github.com/ioccc-src/mkiocccentry
>  * branch            master     -> FETCH_HEAD
> Your local changes to the following files would be overwritten by merge:
>   hostchk.sh
> $ git diff hostchk.sh
> $
> ```
> 
> ***mutters something about GitHub ...***
> 
> When I've fixed that I'll look at the JSON comment that I haven't addressed yet. That way I can look at any problems you brought up.

...been resolved so I'll look at the other thread now though also looking at the code at the same time (well okay going back and forth from terminal and GitHub).
